embedded/devices/src/tca6416.cpp

```diff
@@ -0,0 +1,155 @@
+      this->write_data(TCA_CFG1_REG, 0x01))
+    return -1;
+
+  /*Reset the output registers*/
+  if (this->write_data(TCA_OUT0_REG, 0x01) ||
+      this->write_data(TCA_OUT1_REG, 0x01))
+    return -1;
+
+  /*Reset the input register*/
+  if (this->write_data(TCA_IN0_REG, 0xFF) ||
+      this->write_data(TCA_IN1_REG, 0xFF))
+    return -1;
+
+  return 0;
+}
+
+private:
+int Tca6416::set_dir(bool bank, int pin, uint8_t dir) {
+  uint8_t current_status;
+  uint8_t dirReg;
+  int rc;
+
+  /*return an error if the pin number is invalid*/
+  if (pin >= TCA_NUM_PINS_PER_BANK || pin < 0) return -1;
+
+  /*Make sure you're reading from the right register based on bank*/
+  if (bank) {
+    dirReg = TCA_CFG1_REG;
+  } else {
+    dirReg = TCA_CFG0_REG;
+  }
+  current_status = this->read_from_reg(dirReg);
+
+  /*Write the value of dir to the correct pin*/
+  if (dir) {
+    rc = this->write_data(dirReg, current_status | (1 << pin));
+  } else {
+    rc = this->write_data(dirReg, current_status & ~(1 << pin));
+  }
+  return rc;
+}
+
+private:
+uint8_t Tca6416::get_dir(bool bank, int pin) {
+  uint8_t current_status;
+  uint8_t dirReg;
+
+  /*Return error if pin number invalid*/
+  if (pin >= TCA_NUM_PINS_PER_BANK || pin < 0) return -1;
+
+  /*Decide which register to use based on bank*/
+  if (bank) {
+    dirReg = TCA_CFG1_REG;
+  } else {
+    dirReg = TCA_CFG0_REG;
+  }
+  current_status = read_from_reg(dirReg);
+
+  /* Shift the relevant bit all the way right, then AND with 1 to clear other
+   * bits*/
+  return ((current_status >> pin) & 1);
+}
+
+int Tca6416::begin(const uint8_t directions[]) {
+  int rc;
+  if (!this->isOpen()) {
+    rc = this->open_i2c();
+    if (rc) return rc;
+  }
+  rc = this->clear_settings();
+  if (rc) {
+    std::cerr << "error clearing settings\n";
+    return rc;
+  }
+
+  // set directions for pins in the first bank
+  for (int i = 0; i < TCA_NUM_PINS_PER_BANK; ++i)
+    this->set_dir(0, i, directions[i]);
+  // set directions for pins in the second bank
+  for (int i = 0; i < TCA_NUM_PINS_PER_BANK; ++i)
+    this->set_dir(1, i, directions[TCA_NUM_PINS_PER_BANK + i]);
+
+  // Make sure each pin in bank 0 actually received the correct value
+  for (int i = 0; i < TCA_NUM_PINS_PER_BANK; ++i) {
+    if (this->get_dir(0, i) != directions[i]) {
+      std::cerr << "Error setting direction of pin " << i << "\n";
+      return -EIO;
+    }
+  }
+  // Repeat process for bank 1
+  for (int i = 0; i < TCA_NUM_PINS_PER_BANK; ++i) {
+    if (this->get_dir(1, i) != directions[TCA_NUM_PINS_PER_BANK + i]) {
+      std::cerr << "Error setting direction of pin " << i << "\n";
+      return -EIO;
+    }
+  }
+  return 0;
+}
+
+uint8_t TCA6416::get_state(bool bank, int pin) {
+  uint8_t current_status;
+  uint8_t stateReg;
+
+  if (pin >= TCA_NUM_PINS_PER_BANK || pin < 0) return -EIO;
+
+  // set stateReg to the correct register **Ask Ben if TCA_IN are the right
+  // register to use here
+  if (bank) {
+    stateReg = TCA_IN1_REG;
```

Comment:
Would you be able to make it so that the IN register is only read for input pins, and if we're getting the state of an output pin we read from the OUT reg? You're implementation is good, but I think this will make testing easier